Tkinter problem on darwin

i get this problem on OS X 10.9.5 but not on NixOS.

code of guitest.py

```
#! /usr/bin/env nix-shell
#! nix-shell -i python -p python python27Packages.tkinter

import Tkinter

class simpleapp_tk(Tkinter.Tk):
    def __init__(self,parent):
        Tkinter.Tk.__init__(self,parent)
        self.parent = parent
        self.initialize()

    def initialize(self):
        pass

if __name__ == "__main__":
    app = simpleapp_tk(None)
    app.title('my application')
    app.mainloop()
```

```
Traceback (most recent call last):
  File "./guitest.py", line 16, in <module>
    app = simpleapp_tk(None)
  File "./guitest.py", line 8, in __init__
    Tkinter.Tk.__init__(self,parent)
  File "/nix/store/qp0d2fi7bqsb3vl1qpknc294klq4g9zp-python-2.7.10/lib/python2.7/lib-tk/Tkinter.py", line 1816, in __init__
    self._loadtk()
  File "/nix/store/qp0d2fi7bqsb3vl1qpknc294klq4g9zp-python-2.7.10/lib/python2.7/lib-tk/Tkinter.py", line 1832, in _loadtk
    % (_tkinter.TK_VERSION, tk_version)
RuntimeError: tk.h version (8.6) doesn't match libtk.a version (8.5)
```

same error with `#! nix-shell -i python3 -p python3`

@FRidh thanks.

simple example with python3 also fails:
```
#! /usr/bin/env nix-shell
#! nix-shell -i python3 -p python3

import tkinter
top = tkinter.Tk()
top.mainloop()
```

```
imac:Python davidak$ ./guitest.py
Traceback (most recent call last):
  File "./guitest.py", line 6, in <module>
    import tkinter
  File "/nix/store/gzsbc240vkz13z5wypmlkm2kh1gmw1f6-python3-3.5.2/lib/python3.5/tkinter/__init__.py", line 35, in <module>
    import _tkinter # If this fails your Python may not be configured for Tk
ImportError: No module named '_tkinter'
```

build of python-tkinter-2.7.12 still fails:

```
creating Modules/Setup
creating Modules/Setup.local
creating Makefile
building
running build_ext
sh: file: command not found
building '_tkinter' extension
creating build/temp.macosx-10.7-x86_64-2.7/private
creating build/temp.macosx-10.7-x86_64-2.7/private/var
creating build/temp.macosx-10.7-x86_64-2.7/private/var/folders
creating build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb
creating build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn
creating build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T
creating build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0
creating build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12
creating build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12/Modules
clang -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -DWITH_APPINIT=1 -I/Library/Frameworks/Tcl.framework/Headers -I/Library/Frameworks/Tcl.framework/Versions/Current/PrivateHeaders -I/Library/Frameworks/Tk.framework/Headers -I/Library/Frameworks/Tk.framework/Versions/Current/PrivateHeaders -I/no-such-path/X11R6/include -I/nix/store/n3bdl0kg02d11qpww05bmhx68xq191ld-python-2.7.12/include -I. -IInclude -I./Include -I/nix/store/n3bdl0kg02d11qpww05bmhx68xq191ld-python-2.7.12/include/python2.7 -c /private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12/Modules/_tkinter.c -o build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12/Modules/_tkinter.o -framework Tk
clang-3.7: warning: -framework Tk: 'linker' input unused
clang-3.7: warning: argument unused during compilation: '-fno-strict-overflow'
/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12/Modules/_tkinter.c:1178:38: warning: incompatible pointer types passing 'Py_UNICODE *' (aka 'unsigned int *') to parameter of type 'const Tcl_UniChar *' (aka 'const unsigned short *') [-Wincompatible-pointer-types]
            return Tcl_NewUnicodeObj(inbuf, size);
                                     ^~~~~
/nix/store/xqqg00f9z3p3f1iyasg8qfzr81vjnz1a-tcl-8.6.6/include/tclDecls.h:1103:55: note: passing argument to parameter 'unicode' here
EXTERN Tcl_Obj *        Tcl_NewUnicodeObj(const Tcl_UniChar *unicode,
                                                             ^
1 warning generated.
clang -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -DWITH_APPINIT=1 -I/Library/Frameworks/Tcl.framework/Headers -I/Library/Frameworks/Tcl.framework/Versions/Current/PrivateHeaders -I/Library/Frameworks/Tk.framework/Headers -I/Library/Frameworks/Tk.framework/Versions/Current/PrivateHeaders -I/no-such-path/X11R6/include -I/nix/store/n3bdl0kg02d11qpww05bmhx68xq191ld-python-2.7.12/include -I. -IInclude -I./Include -I/nix/store/n3bdl0kg02d11qpww05bmhx68xq191ld-python-2.7.12/include/python2.7 -c /private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12/Modules/tkappinit.c -o build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12/Modules/tkappinit.o -framework Tk
clang-3.7: warning: -framework Tk: 'linker' input unused
clang-3.7: warning: argument unused during compilation: '-fno-strict-overflow'
creating build/lib.macosx-10.7-x86_64-2.7
clang -bundle -undefined dynamic_lookup build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12/Modules/_tkinter.o build/temp.macosx-10.7-x86_64-2.7/private/var/folders/fb/knpgn__s7_s20d_x1gsllflr0000gn/T/nix-build-python-tkinter-2.7.12.drv-0/Python-2.7.12/Modules/tkappinit.o -L/nix/store/n3bdl0kg02d11qpww05bmhx68xq191ld-python-2.7.12/lib -o build/lib.macosx-10.7-x86_64-2.7/_tkinter.so -framework Tcl -framework Tk
ld: warning: directory not found for option '-L/nix/store/a8gicd4d85bc5j73s6rs601v5fga6yja-xlibs-wrapper/lib'
ld: framework not found Tcl
clang-3.7: error: linker command failed with exit code 1 (use -v to see invocation)

Python build finished, but the necessary bits to build these modules were not found:
_bsddb     _curses    _curses_panel
_sqlite3   _ssl       bsddb185
bz2        dbm        dl
gdbm       imageop    linuxaudiodev
nis        ossaudiodev   readline
spwd       sunaudiodev   zlib
To find the necessary bits, look in setup.py in detect_modules() for the module's name.


Failed to build these modules:
_tkinter

installing
cp: missing destination file operand after '/nix/store/m10dzar03bpksqm6w1wmj7bl6s520r25-python-tkinter-2.7.12/lib/python2.7/site-packages/'
Try 'cp --help' for more information.
builder for ‘/nix/store/jcp1r9xvz8bj2hvwk4qvklzi5szq2pr4-python-tkinter-2.7.12.drv’ failed with exit code 1
error: build of ‘/nix/store/jcp1r9xvz8bj2hvwk4qvklzi5szq2pr4-python-tkinter-2.7.12.drv’ failed
```

on master (nixpkgs-17.03pre93141.e3deccc)

--- a/pkgs/development/interpreters/python/cpython/2.7/default.nix
+++ b/pkgs/development/interpreters/python/cpython/2.7/default.nix
@@ -60,6 +60,8 @@ let
 
       ./properly-detect-curses.patch
 
+    ] ++ optionals (x11Support && stdenv.isDarwin) [
+      ./use-correct-tcl-tk-on-darwin.patch
     ] ++ optionals stdenv.isLinux [
 
       # Disable the use of ldconfig in ctypes.util.find_library (since

--- a/pkgs/development/interpreters/python/cpython/2.7/use-correct-tcl-tk-on-darwin.patch
+++ b/pkgs/development/interpreters/python/cpython/2.7/use-correct-tcl-tk-on-darwin.patch
@@ -1,0 +1,48 @@
+diff --git a/setup.py b/setup.py
+index 2779658..902d0eb 100644
+--- a/setup.py
++++ b/setup.py
+@@ -1699,9 +1699,6 @@ class PyBuildExt(build_ext):
+         # Rather than complicate the code below, detecting and building
+         # AquaTk is a separate method. Only one Tkinter will be built on
+         # Darwin - either AquaTk, if it is found, or X11 based Tk.
+-        if (host_platform == 'darwin' and
+-            self.detect_tkinter_darwin(inc_dirs, lib_dirs)):
+-            return
+
+         # Assume we haven't found any of the libraries or include files
+         # The versions with dots are used on Unix, and the versions without
+@@ -1747,22 +1744,6 @@ class PyBuildExt(build_ext):
+             if dir not in include_dirs:
+                 include_dirs.append(dir)
+
+-        # Check for various platform-specific directories
+-        if host_platform == 'sunos5':
+-            include_dirs.append('/usr/openwin/include')
+-            added_lib_dirs.append('/usr/openwin/lib')
+-        elif os.path.exists('/usr/X11R6/include'):
+-            include_dirs.append('/usr/X11R6/include')
+-            added_lib_dirs.append('/usr/X11R6/lib64')
+-            added_lib_dirs.append('/usr/X11R6/lib')
+-        elif os.path.exists('/usr/X11R5/include'):
+-            include_dirs.append('/usr/X11R5/include')
+-            added_lib_dirs.append('/usr/X11R5/lib')
+-        else:
+-            # Assume default location for X11
+-            include_dirs.append('/usr/X11/include')
+-            added_lib_dirs.append('/usr/X11/lib')
+-
+         # If Cygwin, then verify that X is installed before proceeding
+         if host_platform == 'cygwin':
+             x11_inc = find_file('X11/Xlib.h', [], include_dirs)
+@@ -1786,10 +1767,6 @@ class PyBuildExt(build_ext):
+         if host_platform in ['aix3', 'aix4']:
+             libs.append('ld')
+
+-        # Finally, link with the X11 libraries (not appropriate on cygwin)
+-        if host_platform != "cygwin":
+-            libs.append('X11')
+-
+         ext = Extension('_tkinter', ['_tkinter.c', 'tkappinit.c'],
+                         define_macros=[('WITH_APPINIT', 1)] + defs,
+                         include_dirs = include_dirs,

--- a/pkgs/development/interpreters/python/cpython/3.4/default.nix
+++ b/pkgs/development/interpreters/python/cpython/3.4/default.nix
@@ -67,6 +67,8 @@ in stdenv.mkDerivation {
   patches = [
     ./no-ldconfig.patch
     ./ld_library_path.patch
+  ] ++ optionals (x11Support && stdenv.isDarwin) [
+    ./use-correct-tcl-tk-on-darwin.patch
   ];
 
   postPatch = ''

--- a/pkgs/development/interpreters/python/cpython/3.4/use-correct-tcl-tk-on-darwin.patch
+++ b/pkgs/development/interpreters/python/cpython/3.4/use-correct-tcl-tk-on-darwin.patch
@@ -1,0 +1,48 @@
+diff --git a/setup.py b/setup.py
+index 2779658..902d0eb 100644
+--- a/setup.py
++++ b/setup.py
+@@ -1699,9 +1699,6 @@ class PyBuildExt(build_ext):
+         # Rather than complicate the code below, detecting and building
+         # AquaTk is a separate method. Only one Tkinter will be built on
+         # Darwin - either AquaTk, if it is found, or X11 based Tk.
+-        if (host_platform == 'darwin' and
+-            self.detect_tkinter_darwin(inc_dirs, lib_dirs)):
+-            return
+
+         # Assume we haven't found any of the libraries or include files
+         # The versions with dots are used on Unix, and the versions without
+@@ -1747,22 +1744,6 @@ class PyBuildExt(build_ext):
+             if dir not in include_dirs:
+                 include_dirs.append(dir)
+
+-        # Check for various platform-specific directories
+-        if host_platform == 'sunos5':
+-            include_dirs.append('/usr/openwin/include')
+-            added_lib_dirs.append('/usr/openwin/lib')
+-        elif os.path.exists('/usr/X11R6/include'):
+-            include_dirs.append('/usr/X11R6/include')
+-            added_lib_dirs.append('/usr/X11R6/lib64')
+-            added_lib_dirs.append('/usr/X11R6/lib')
+-        elif os.path.exists('/usr/X11R5/include'):
+-            include_dirs.append('/usr/X11R5/include')
+-            added_lib_dirs.append('/usr/X11R5/lib')
+-        else:
+-            # Assume default location for X11
+-            include_dirs.append('/usr/X11/include')
+-            added_lib_dirs.append('/usr/X11/lib')
+-
+         # If Cygwin, then verify that X is installed before proceeding
+         if host_platform == 'cygwin':
+             x11_inc = find_file('X11/Xlib.h', [], include_dirs)
+@@ -1786,10 +1767,6 @@ class PyBuildExt(build_ext):
+         if host_platform in ['aix3', 'aix4']:
+             libs.append('ld')
+
+-        # Finally, link with the X11 libraries (not appropriate on cygwin)
+-        if host_platform != "cygwin":
+-            libs.append('X11')
+-
+         ext = Extension('_tkinter', ['_tkinter.c', 'tkappinit.c'],
+                         define_macros=[('WITH_APPINIT', 1)] + defs,
+                         include_dirs = include_dirs,

--- a/pkgs/development/interpreters/python/cpython/3.5/default.nix
+++ b/pkgs/development/interpreters/python/cpython/3.5/default.nix
@@ -67,6 +67,8 @@ in stdenv.mkDerivation {
   patches = [
     ./no-ldconfig.patch
     ./ld_library_path.patch
+  ] ++ optionals (x11Support && stdenv.isDarwin) [
+    ./use-correct-tcl-tk-on-darwin.patch
   ];
 
   postPatch = ''

--- a/pkgs/development/interpreters/python/cpython/3.5/use-correct-tcl-tk-on-darwin.patch
+++ b/pkgs/development/interpreters/python/cpython/3.5/use-correct-tcl-tk-on-darwin.patch
@@ -1,0 +1,48 @@
+diff --git a/setup.py b/setup.py
+index 2779658..902d0eb 100644
+--- a/setup.py
++++ b/setup.py
+@@ -1699,9 +1699,6 @@ class PyBuildExt(build_ext):
+         # Rather than complicate the code below, detecting and building
+         # AquaTk is a separate method. Only one Tkinter will be built on
+         # Darwin - either AquaTk, if it is found, or X11 based Tk.
+-        if (host_platform == 'darwin' and
+-            self.detect_tkinter_darwin(inc_dirs, lib_dirs)):
+-            return
+
+         # Assume we haven't found any of the libraries or include files
+         # The versions with dots are used on Unix, and the versions without
+@@ -1747,22 +1744,6 @@ class PyBuildExt(build_ext):
+             if dir not in include_dirs:
+                 include_dirs.append(dir)
+
+-        # Check for various platform-specific directories
+-        if host_platform == 'sunos5':
+-            include_dirs.append('/usr/openwin/include')
+-            added_lib_dirs.append('/usr/openwin/lib')
+-        elif os.path.exists('/usr/X11R6/include'):
+-            include_dirs.append('/usr/X11R6/include')
+-            added_lib_dirs.append('/usr/X11R6/lib64')
+-            added_lib_dirs.append('/usr/X11R6/lib')
+-        elif os.path.exists('/usr/X11R5/include'):
+-            include_dirs.append('/usr/X11R5/include')
+-            added_lib_dirs.append('/usr/X11R5/lib')
+-        else:
+-            # Assume default location for X11
+-            include_dirs.append('/usr/X11/include')
+-            added_lib_dirs.append('/usr/X11/lib')
+-
+         # If Cygwin, then verify that X is installed before proceeding
+         if host_platform == 'cygwin':
+             x11_inc = find_file('X11/Xlib.h', [], include_dirs)
+@@ -1786,10 +1767,6 @@ class PyBuildExt(build_ext):
+         if host_platform in ['aix3', 'aix4']:
+             libs.append('ld')
+
+-        # Finally, link with the X11 libraries (not appropriate on cygwin)
+-        if host_platform != "cygwin":
+-            libs.append('X11')
+-
+         ext = Extension('_tkinter', ['_tkinter.c', 'tkappinit.c'],
+                         define_macros=[('WITH_APPINIT', 1)] + defs,
+                         include_dirs = include_dirs,

--- a/pkgs/development/interpreters/python/cpython/3.6/default.nix
+++ b/pkgs/development/interpreters/python/cpython/3.6/default.nix
@@ -69,6 +69,8 @@ in stdenv.mkDerivation {
 
   patches = [
     ./no-ldconfig.patch
+  ] ++ optionals (x11Support && stdenv.isDarwin) [
+    ./use-correct-tcl-tk-on-darwin.patch
   ];
 
   postPatch = ''

--- a/pkgs/development/interpreters/python/cpython/3.6/use-correct-tcl-tk-on-darwin.patch
+++ b/pkgs/development/interpreters/python/cpython/3.6/use-correct-tcl-tk-on-darwin.patch
@@ -1,0 +1,48 @@
+diff --git a/setup.py b/setup.py
+index 2779658..902d0eb 100644
+--- a/setup.py
++++ b/setup.py
+@@ -1699,9 +1699,6 @@ class PyBuildExt(build_ext):
+         # Rather than complicate the code below, detecting and building
+         # AquaTk is a separate method. Only one Tkinter will be built on
+         # Darwin - either AquaTk, if it is found, or X11 based Tk.
+-        if (host_platform == 'darwin' and
+-            self.detect_tkinter_darwin(inc_dirs, lib_dirs)):
+-            return
+
+         # Assume we haven't found any of the libraries or include files
+         # The versions with dots are used on Unix, and the versions without
+@@ -1747,22 +1744,6 @@ class PyBuildExt(build_ext):
+             if dir not in include_dirs:
+                 include_dirs.append(dir)
+
+-        # Check for various platform-specific directories
+-        if host_platform == 'sunos5':
+-            include_dirs.append('/usr/openwin/include')
+-            added_lib_dirs.append('/usr/openwin/lib')
+-        elif os.path.exists('/usr/X11R6/include'):
+-            include_dirs.append('/usr/X11R6/include')
+-            added_lib_dirs.append('/usr/X11R6/lib64')
+-            added_lib_dirs.append('/usr/X11R6/lib')
+-        elif os.path.exists('/usr/X11R5/include'):
+-            include_dirs.append('/usr/X11R5/include')
+-            added_lib_dirs.append('/usr/X11R5/lib')
+-        else:
+-            # Assume default location for X11
+-            include_dirs.append('/usr/X11/include')
+-            added_lib_dirs.append('/usr/X11/lib')
+-
+         # If Cygwin, then verify that X is installed before proceeding
+         if host_platform == 'cygwin':
+             x11_inc = find_file('X11/Xlib.h', [], include_dirs)
+@@ -1786,10 +1767,6 @@ class PyBuildExt(build_ext):
+         if host_platform in ['aix3', 'aix4']:
+             libs.append('ld')
+
+-        # Finally, link with the X11 libraries (not appropriate on cygwin)
+-        if host_platform != "cygwin":
+-            libs.append('X11')
+-
+         ext = Extension('_tkinter', ['_tkinter.c', 'tkappinit.c'],
+                         define_macros=[('WITH_APPINIT', 1)] + defs,
+                         include_dirs = include_dirs,

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -16394,15 +16394,14 @@ in {
       # Move the tkinter module
       mkdir -p $out/${py.sitePackages}
       mv lib/${py.libPrefix}/lib-dynload/_tkinter* $out/${py.sitePackages}/
+    '' + stdenv.lib.optionalString (!stdenv.isDarwin) ''
       # Update the rpath to point to python without x11Support
       old_rpath=$(patchelf --print-rpath $out/${py.sitePackages}/_tkinter*)
       new_rpath=$(sed "s#${py}#${python}#g" <<< "$old_rpath" )
       patchelf --set-rpath $new_rpath $out/${py.sitePackages}/_tkinter*
     '';
 
-    meta = py.meta // {
-      platforms = platforms.linux;
-    };
+    meta = py.meta;
   };
 
   tlslite = buildPythonPackage rec {

